Gpio command controller (#1251)
ros-controls/ros2_controllers: gpio_controllers/doc/userdoc.rst

.. _gpio_controllers_userdoc:

gpio_controllers
=====================

> This is a collection of controllers for hardware interfaces of type GPIO (``<gpio>`` tag in the URDF).

Reviewer: ```suggestion
A controller for hardware interfaces of type GPIO (``<gpio>`` tag in the URDF) to be able to command the configured GPIOs and as well to publish their state
```

Reviewer 2: I think this was OK: gpio_controllers follow the structure of other collection packages we have (effort_controller), where only a single controller is implemented yet (gpio_controllers/gpio_command_controller, effort_controllers/JointGroupEffortController)

Reviewer: Gotcha! 👍🏾

.. _gpio_controllers_userdoc:

gpio_controllers
=====================

This is a collection of controllers for hardware interfaces of type GPIO (``<gpio>`` tag in the URDF).

gpio_command_controller
-----------------------------
gpio_command_controller let the user expose command interfaces of given GPIO interfaces and publishes state interfaces of the configured command interfaces.

Description of controller's interfaces
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- ``/<controller_name>/gpio_states`` [``control_msgs/msg/DynamicJointState``]: Publishes all state interfaces of the given GPIO interfaces.
- ``/<controller_name>/commands`` [``control_msgs/msg/DynamicJointState``]:  A subscriber for configured command interfaces.

Parameters
^^^^^^^^^^^^^^^^^^^^^^^^

> This controller uses the `generate_parameter_library <https://github.com/PickNikRobotics/generate_parameter_library>`_ to handle its parameters. The parameter `definition file located in the src folder <https://github.com/ros-controls/ros2_controllers/blob/{REPOS_FILE_BRANCH}/gpio_controllers_/src/gpio_command_controller_parameters.yaml>`_ contains descriptions for all the parameters used by the controller.

Reviewer 2: ```suggestion
This controller uses the `generate_parameter_library <https://github.com/PickNikRobotics/generate_parameter_library>`_ to handle its parameters. The parameter `definition file located in the src folder <https://github.com/ros-controls/ros2_controllers/blob/{REPOS_FILE_BRANCH}/gpio_controllers/src/gpio_command_controller_parameters.yaml>`_ contains descriptions for all the parameters used by the controller.
```

.. generate_parameter_library_details::
  ../src/gpio_command_controller_parameters.yaml

The controller expects at least one GPIO interface and the corresponding command interface names or state interface. However, these Command and State interfaces are optional. The controller behaves as a broadcaster when no Command Interface is present, thereby publishing the configured GPIO state interfaces if set, else the one present in the URDF.

.. note::

  When no state interface is provided in the param file, the controller will try to use state_interfaces from ros2_control's config placed in the URDF for configured gpio interfaces.

>   However, Command interfaces will not be configured based on the available URDF setup.

Reviewer 2: ```suggestion
  However, command interfaces will not be configured based on the available URDF setup.
```

gpio_command_controller
-----------------------------
gpio_command_controller let the user expose command interfaces of given GPIO interfaces and publishes state interfaces of the configured command interfaces.

Description of controller's interfaces
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- ``/<controller_name>/gpio_states`` [``control_msgs/msg/DynamicJointState``]: Publishes all state interfaces of the given GPIO interfaces.
- ``/<controller_name>/commands`` [``control_msgs/msg/DynamicJointState``]:  A subscriber for configured command interfaces.

Parameters
^^^^^^^^^^^^^^^^^^^^^^^^

This controller uses the `generate_parameter_library <https://github.com/PickNikRobotics/generate_parameter_library>`_ to handle its parameters. The parameter `definition file located in the src folder <https://github.com/ros-controls/ros2_controllers/blob/{REPOS_FILE_BRANCH}/gpio_controllers_/src/gpio_command_controller_parameters.yaml>`_ contains descriptions for all the parameters used by the controller.

.. generate_parameter_library_details::
  ../src/gpio_command_controller_parameters.yaml

> The controller expects at least one GPIO interface and the corresponding command interface names or state interface.
> Command and state interfaces are optional.

Reviewer: ```suggestion
The controller expects at least one GPIO interface and the corresponding command interface names or state interface. However, these Command and State interfaces are optional. The controller behaves as a broadcaster when no Command Interface is present, thereby publishing the configured GPIO state interfaces if set, else the one present in the URDF.
```
